libkbfs/disk_block_cache.go

```diff
@@ -0,0 +1,316 @@
+// Copyright 2017 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libkbfs
+
+import (
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/keybase/client/go/logger"
+	"github.com/keybase/kbfs/kbfsblock"
+	"github.com/keybase/kbfs/kbfscrypto"
+	"github.com/keybase/kbfs/tlf"
+	"github.com/syndtr/goleveldb/leveldb"
+	"github.com/syndtr/goleveldb/leveldb/storage"
+	"github.com/syndtr/goleveldb/leveldb/util"
+	"golang.org/x/net/context"
+)
+
+const (
+	// 10 GB maximum storage by default
+	defaultDiskBlockCacheMaxBytes uint64 = 10 * (1 << 30)
+	evictionConsiderationFactor   uint   = 3
+	blockDbFilename               string = "diskCacheBlocks.leveldb"
+	lruDbFilename                 string = "diskCacheLRU.leveldb"
+)
+
+// diskBlockCacheEntry packages an encoded block and serverHalf into one data
+// structure, allowing us to encode it as one set of bytes.
+type diskBlockCacheEntry struct {
+	Buf        []byte
+	ServerHalf kbfscrypto.BlockCryptKeyServerHalf
+}
+
+// diskBlockCacheConfig specifies the interfaces that a DiskBlockCacheStandard
+// needs to perform its functions. This adheres to the standard libkbfs Config
+// API.
+type diskBlockCacheConfig interface {
```

Comment:
As stated in my other comment, I disagree with never inlining arguments. The "flexibility in adding new `Config` parameters with minimal code change" is optimizing for ease of writing at the expense of ease of reading, which is the wrong thing. I'd rather see a call to `NewFoo(param1, param2, param3)` or `NewFoo(FooConfigStruct{Param1: ..., ParamN: ...})` than `NewFoo(config)`. If I'm adding a new type, I _want_ to see all the places that have to be changed to accommodate that new type.

Also, I don't think this is getting rid of more boilerplate than it's creating. Now I see one-off interfaces like `codecGetter`, etc., which only return a single type. There's no difference between that and the original `Codec` interface -- the extra level of indirection can be done by providing a switchable `Codec` interface itself, and in 99% of tests, that extra level of indirection is never used anyway.